test/lambda_calc/lambda_calc.ml

```diff
@@ -0,0 +1,109 @@
+(* <variable> = 'a'-'z' ; ('a' - 'z' or 'A'-'Z' or '0' - '9')^* *)
+type variable = string
+
+(* <lambda-exppr> = 
+      <variable> | lambda <variable> <lambda_expr> | <lambda_expr> <lambda_expr>
+   *)
+type lambda_expr = 
+    Variable of variable 
+  | Lambda of variable * lambda_expr 
+  | Application of lambda_expr * lambda_expr
+
+type tokens = LPAREN | RPAREN | LAMBDA | DOT | CHAR of char
+
+module Token = 
+struct 
+  type t = tokens
+  let equal (x : t) (y : t) : bool = x = y
+end
+
+module TokenParserCombinator = ParserCombinator.Make(ParserM.Make(Token))
+open TokenParserCombinator
+
+let check_char p = function 
+| CHAR c -> p c
+| _ -> false
+
+let is_lowercase c = Char.code 'a' <= Char.code c && Char.code c <= Char.code 'z'
+let lowercase : token parser_m = sat (check_char is_lowercase)
+let is_uppercase c = Char.code 'A' <= Char.code c && Char.code c <= Char.code 'Z'
+let uppercase : token parser_m = sat (check_char is_uppercase)
+let is_numeric c = Char.code '0' <= Char.code c && Char.code c <= Char.code '9'
+let numeric : token parser_m = sat (check_char is_numeric)
+
+let alphanum = lowercase ++ uppercase ++ numeric
+
+let variable = 
+  let* first = lowercase in 
+  let* rest = star alphanum in 
+
+  let variable_name_list = first :: rest in 
+  let get_chars = function 
+  | CHAR c -> c 
+  | _ -> failwith "all tokens should be CHARs" in 
+
+  let chars = List.map get_chars variable_name_list in 
+  let b = Buffer.create 0 in 
+  List.iter (Buffer.add_char b) chars;
+
+  return (Variable (Buffer.contents b))
+
+let lambda : token parser_m = sat (Token.equal LAMBDA) 
+let dot : token parser_m = sat (Token.equal DOT) 
+let lparen : token parser_m = sat (Token.equal LPAREN)
+let rparen : token parser_m = sat (Token.equal RPAREN)
+
+let rec lambda_expr ()  = variable ++ lambda_fn () ++ application ()
+and lambda_fn () = 
+  let* _ = lambda in 
+  let* variable_name = variable in 
```

Comment:
You should probably define `variable_str` such that you can parse that here and `variable` is `variable_str >>= Variable`

Reply:
that obviates the need for the `assert false` type beat below